plugin/caddyfile/processor.go

```diff
@@ -1,43 +1,39 @@
-package plugin
+package caddyfile
 
 import (
 	"bytes"
 	"fmt"
-	"log"
-	"sort"
 	"strings"
 
-	"github.com/caddyserver/caddy"
-	"github.com/caddyserver/caddy/caddyfile"
-
-	_ "github.com/caddyserver/caddy/caddyhttp" // plug in the HTTP server type
+	"github.com/caddyserver/caddy/v2/caddyconfig"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
 )
 
-// ProcessCaddyfile validate and removes wrong server blocks from caddyfile
-func ProcessCaddyfile(caddyfileContent []byte) []byte {
-	serverBlocks, err := caddyfile.Parse("", bytes.NewReader(caddyfileContent), nil)
+// Process caddyfile and removes wrong server blocks
+func Process(caddyfileContent []byte) ([]byte, []byte) {
+	logsBuffer := bytes.Buffer{}
+
+	serverBlocks, err := caddyfile.Parse("", caddyfileContent)
 
 	if err != nil {
-		log.Printf("[ERROR] Error parsing caddyfile:%s\n", err)
+		logsBuffer.WriteString(fmt.Sprintf("[ERROR] Error parsing caddyfile: %s\n", err.Error()))
```

Comment:
In v2 you can use `zap` for logging. Here's a note from the docs: https://caddyserver.com/docs/extending-caddy#logs

Reply:
I played a bit with zap on another branch. But I have a specific requirement here and I didn't find how to do with it.

Because we do polling and monitor all events from docker, we generate caddyfile several times but if it is the same as the last generated one, we discard the logs and new caddyfile.

Thus, the logsBuffer is there. To store logs without actually printing them.

Do you have any suggestions?

I think this can be done in another PR later.

Reply:
Hmm, maybe make like a TempLog struct, something like `Level, Msg` and append to it, then later emit all the logs with zap at once?

You could override the log level for each one instead of discarding, setting them all to `debug`. That way, if a user sets the `debug` global option, they'll get a bit more info that might help with figuring out why nothing happened when they spun up a new container 😄

Reply:
I wonder if a better way to do this is to implement a log writer that can buffer/discard logs as it sees fit?

Reply:
Agree, I will merge without that first.

Reply:
https://github.com/lucaslorentz/caddy-docker-proxy/issues/139